rfd/0164-scoped-rbac.md

```diff
@@ -0,0 +1,956 @@
+
+```yaml
+kind: access_list
+metadata:
+  name: access-to-default
+spec:
+  grants: 
+    roles: [access]
+    traits:
+      'internal.logins' : 'root'
+  scope:  '/k8s/namespaces/default`
+  members:
+    - bob@example.com
+```
+
+### Scoped search-based access requests
+
+Search-based access requests let users to search and request access to individual resources. Here is a standard requester role:
+
+```yaml
+# requester.yaml
+kind: role
+version: v5
+metadata:
+  name: requester
+spec:
+  allow:
+    request:
+      search_as_roles:
+        - access
+```
+
+Here is a standard reviewer role:
+
+```yaml
+# reviewer.yaml
+kind: role
+version: v5
+metadata:
+  name: reviewer
+spec:
+  allow:
+    review_requests:
+      roles:
+        - access
+      preview_as_roles:
+        - access
+```
+
+Without changing those roles, we can assign both requester and reviewer roles in a specific scope with access list:
+
+```yaml
+kind: access_list
+metadata:
+  name: access-to-default
+spec:
+  grants: 
+    roles: [requester, reviewer]
+  scope:  '/dev`
+  members:
+    - bob@example.com
+ - alice@example.com
+```
+
+In this case, `bob@example.com` and `alice@example.com` will get an ability to search, request and review requests, but only in the scope of any resource in `/dev` resource group.
+
+Customers frequently ask a question of how to scale this with multiple teams, 
+with this approach, we’d have to create an access list for each individual team. 
+
+Previously we’ve been recommending to use role templates. 
+However, new access lists integration mirrors any group hierarchy in identity providers, 
+so there is no need to use templates - Teleport will create access lists and keep members up to date.
+
+The only thing we are missing is to let customers specify the scope when importing Okta groups or apps as access lists. 
+For example, access list for Okta group `devs` can automatically have scope `/dev`
+
+Additionally, one access list can be a member of another access list. Let’s review a case when we have a group devs that needs access to both staging and production.
+
+Let’s create access list for Alice and Bob, this special access list does no grants and scopes, we are going to use it to keep list of our developers:
+
+```yaml
+kind: access_list
+metadata:
+  name: dev-team
+spec:
+  members:
+    - bob@example.com
+    - alice@example.com
+---
+kind: access_list
+metadata:
+  name: access-to-prod
+spec:
+  grants: 
+    roles: [requester, reviewer]
+  scope:  '/dev`
+  member_lists:
+    - dev-team
+---
+kind: access_list
+metadata:
+  name: access-to-stage
+spec:
+  grants: 
+    roles: [requester, reviewer]
+  scope:  '/prod`
+  member_lists:
+    - dev-team
+```
+
+**Note:** We have to make sure that the child access list is a strict subset of the roles, traits and scopes of it's parent or has no scopes or grants at all.
+
+### Scoped Impersonation
```

Comment:
Perhaps my thoughts here are a little tangential to this RFD, but part of me has been wondering whether impersonation is something we need to support externally long-term. Historically, it has been a useful tool in producing long-lived credentials for a Teleport user to act as a service account from machine to machine environments like CI/CD, but Machine ID is now the golden path for this. Recent changes (e.g requiring MFA for Admin RPCs) has made impersonation even less useful with tools like our Terraform provider or `tctl`.

Perhaps the only usecase I can see into the future is for those trying to debug their RBAC - and assuming a user is the best way of exploring their access. However, we could potentially build out other tools that make this easier (e.g a CLI or UI tool that allows the RBAC configuration to be tested for a given user/roleset)

Reply:
I'd keep this out of scope of this RFD. We likely have a lot of customers leveraging impersonation. I think it would work as is with the new design, and you can plan transition to a better model when convenient for Machine ID.